test/states/test.Postroll.js

```diff
@@ -0,0 +1,152 @@
+import QUnit from 'qunit';
+
+import {Postroll} from '../../src/states.js';
+import adBreak from '../../src/adBreak.js';
+
+/*
+ * These tests are intended to be isolated unit tests for one state with all
+ * other modules mocked.
+ */
+QUnit.module('Postroll', {
+  beforeEach: function() {
+    this.events = [];
+
+    this.player = {
+      ads: {
+        settings: {},
+        debug: () => {},
+        inAdBreak: () => false
+      },
+      addClass: () => {},
+      removeClass: () => {},
+      setTimeout: () => {},
+      trigger: (event) => {
+        this.events.push(event);
+      },
+      clearTimeout: () => {}
+    };
+
+    this.postroll = new Postroll(this.player);
+
+    this.postroll.transitionTo = (newState) => {
+      this.transitionTo = newState.name;
+    };
+
+    this.adBreakStartStub = sinon.stub(adBreak, 'start');
+    this.adBreakEndStub = sinon.stub(adBreak, 'end');
+  },
+
+  afterEach() {
+    this.adBreakStartStub.restore();
+    this.adBreakEndStub.restore();
+  }
+});
+
+QUnit.test('sets _contentEnding on init', function(assert) {
+  this.postroll.init(this.player);
+  assert.equal(this.player.ads._contentEnding, true, 'content is ending');
+});
+
+QUnit.test('startLinearAdMode starts ad break', function(assert) {
+  this.postroll.init(this.player);
+  this.postroll.startLinearAdMode();
+  assert.equal(this.adBreakStartStub.callCount, 1, 'ad break started');
+});
+
+QUnit.test('removes ad loading class on ad started', function(assert) {
+  this.player.removeClass = sinon.spy();
+  this.postroll.init(this.player);
+  this.postroll.onAdStarted(this.player);
+  assert.ok(this.player.removeClass.calledWith('vjs-ad-loading'));
+});
+
+QUnit.test('ends linear ad mode & ended event on ads error', function(assert) {
+  this.player.ads.endLinearAdMode = sinon.spy();
+
+  this.postroll.init(this.player);
+  this.player.ads.inAdBreak = () => true;
+  this.postroll.onAdsError(this.player);
+  assert.equal(this.player.ads.endLinearAdMode.callCount, 1, 'linear ad mode ended');
+  assert.equal(this.events[0], 'ended', 'saw ended event');
+});
+
+QUnit.test('no endLinearAdMode on adserror if not in ad break', function(assert) {
+  this.player.ads.endLinearAdMode = sinon.spy();
+
+  this.postroll.init(this.player);
+  this.player.ads.inAdBreak = () => false;
+  this.postroll.onAdsError(this.player);
+  assert.equal(this.player.ads.endLinearAdMode.callCount, 0, 'linear ad mode ended');
```

Comment:
does ended get called?

Reply:
might as well assert not

Reply:
sorry the other way, we assert that ended is still called